File: ros2_ws/src/px4_ros_com/src/examples/tests/4d_trajectory_timeopt.py
"""
uav_pso_4d_timeopt.py
Segunda versión: optimiza posiciones (x,y,z) y tiempos (delta-t) por intervalo.

Concepto:
- Cada UAV tiene T posiciones (incluye inicio y fin) y T-1 intervalos de tiempo.
- Los intervalos de tiempo son variables y positivas; los tiempos de cada paso son acumulados.
- Las velocidades se calculan como distancia / delta_t y se penalizan si exceden max_speed.
- La detección de colisiones usa interpolación lineal para evaluar la posición de otros UAVs en tiempos arbitrarios.

Uso:
    python 4d_trajectory_timeopt.py

Autor: José Antonio García Campanario
Fecha: 2025-10-31
"""

import numpy as np
import math
import os
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
import matplotlib.cm as cm

# ----- Escenario (copiado/adaptado) -----
n_uavs = 3
T = 25
dt_base = 1.0           # tiempo base (escala para inicialización)
max_speed = 3.0         # unidades de distancia por segundo
min_sep = 2.0
space_bounds = np.array([[0, 40], [0, 40], [0.5, 10]])

starts = np.array([[2, 2, 3],
                   [2, 8, 3],
                   [2, 14, 3]], dtype=float)
goals  = np.array([[35, 35, 4],
                   [35, 30, 6],
                   [35, 25, 3]], dtype=float)

# ----- Helpers -----

def clamp_pos(pos):
    for dim in range(3):
        pos[..., dim] = np.clip(pos[..., dim], space_bounds[dim,0], space_bounds[dim,1])
    return pos


def cumulative_times_from_deltas(deltas):
    """deltas: (n_uavs, T-1) -> times: (n_uavs, T)
    times[:,0] = 0
    times[:,i] = sum_{k< i} deltas[:,k]
    """
    n, m = deltas.shape
    times = np.zeros((n, m+1))
    times[:,1:] = np.cumsum(deltas, axis=1)
    return times


def interp_pos_at_time(traj, times, t_query):
    """Interpolación lineal en una trayectoria de un UAV.
    traj: (T,3), times: (T,), t_query: scalar
    Devuelve posición interpolada. Si t_query fuera de rango, devuelve endpoint.
    """
    if t_query <= times[0]:
        return traj[0]
    if t_query >= times[-1]:
        return traj[-1]
    # localizar segmento
    idx = np.searchsorted(times, t_query) - 1
    t0 = times[idx]
    t1 = times[idx+1]
    p0 = traj[idx]
    p1 = traj[idx+1]
    if t1 == t0:
        return p0
    alpha = (t_query - t0) / (t1 - t0)
    return (1-alpha) * p0 + alpha * p1

# ----- Cost components (accept time deltas) -----

def path_length(trajectories):
    diffs = np.diff(trajectories, axis=1)
    return np.sum(np.linalg.norm(diffs, axis=2))


def speed_penalty_timeopt(trajectories, deltas):
    """trajectories: (n_uavs,T,3), deltas: (n_uavs,T-1)
    Penaliza velocidades que exceden max_speed (vel = dist / delta)
    """
    pen = 0.0
    diffs = np.diff(trajectories, axis=1)
    dists = np.linalg.norm(diffs, axis=2)  # (n_uavs, T-1)
    speeds = dists / deltas
    excess = np.maximum(0.0, speeds - max_speed)
    pen += np.sum(excess**2) * 100.0
    # penalizar delta muy pequeños (para evitar t->0)
    pen += np.sum(np.maximum(0.0, 0.05 - deltas)**2) * 1000.0
    return pen


def collision_penalty_timeopt(trajectories, deltas):
    """Considera tiempos diferentes por UAV y usa interpolación para chequear colisiones.
    Para cada UAV u y cada timestamp t_u[i], evalúa la posición del otro UAV v interpolada en ese tiempo.
    """
    pen = 0.0
    times = cumulative_times_from_deltas(deltas)
    for u in range(n_uavs):
        for i in range(T):
            t_u = times[u, i]
            p_u = trajectories[u, i]
            for v in range(u+1, n_uavs):
                p_v = interp_pos_at_time(trajectories[v], times[v], t_u)
                d = np.linalg.norm(p_u - p_v)
                if d < min_sep:
                    pen += (min_sep - d)**2 * 200.0
    return pen


def boundary_penalty(trajectories):
    pen = 0.0
    for dim in range(3):
        low = space_bounds[dim,0]
        high = space_bounds[dim,1]
        below = np.minimum(0.0, trajectories[...,dim] - low)
        above = np.maximum(0.0, trajectories[...,dim] - high)
        pen += np.sum(below**2 + above**2) * 100.0
    return pen

# objective: takes flat vector with positions and deltas

def objective_timeopt(flat_vec):
    # flat_vec layout: positions (n_uavs*T*3) then deltas (n_uavs*(T-1))
    pos_len = n_uavs * T * 3
    del_len = n_uavs * (T-1)
    pos_flat = flat_vec[:pos_len]
    del_flat = flat_vec[pos_len: pos_len + del_len]
    traj = pos_flat.reshape((n_uavs, T, 3))
    deltas = del_flat.reshape((n_uavs, T-1))
    # anchor start and end softly via penalty
    start_err = np.sum((traj[:,0,:] - starts)**2) * 1000.0
    goal_err  = np.sum((traj[:,-1,:] - goals)**2) * 1000.0
    L = path_length(traj)
    sp = speed_penalty_timeopt(traj, deltas)
    coll = collision_penalty_timeopt(traj, deltas)
    bnd = boundary_penalty(traj)
    # smoothness on positions and time deltas
    pos_smooth = np.sum(np.linalg.norm(np.diff(traj, n=2, axis=1), axis=2)) * 10.0
    time_smooth = np.sum(np.abs(np.diff(deltas, axis=1))) * 50.0
    cost = L + pos_smooth + time_smooth + sp + coll + bnd + start_err + goal_err
    return cost

# ----- PSO for time-optimized trajectories -----

def run_pso_timeopt(n_particles=300, iters=300, w=0.6, c1=1.8, c2=1.8, seed=None):
    if seed is not None:
        np.random.seed(seed)
    pos_dim = n_uavs * T * 3
    del_dim = n_uavs * (T-1)
    dim = pos_dim + del_dim

    particles = np.zeros((n_particles, dim))
    velocities = np.zeros_like(particles)
    pbest = np.zeros_like(particles)
    pbest_cost = np.full(n_particles, np.inf)
    gbest = None
    gbest_cost = np.inf

    # Initialize particles: positions along straight line + small noise; deltas = dt_base + noise
    for p in range(n_particles):
        traj = np.zeros((n_uavs, T, 3))
        deltas = np.zeros((n_uavs, T-1))
        for i in range(n_uavs):
            for k in range(3):
                traj[i, :, k] = np.linspace(starts[i,k], goals[i,k], T)
            traj[i] += np.random.normal(scale=0.6, size=traj[i].shape)
            # deltas around dt_base
            deltas[i] = np.clip(dt_base + np.random.normal(scale=0.1, size=(T-1,)), 0.05, 10.0)
        traj = clamp_pos(traj)
        particles[p, :pos_dim] = traj.reshape(-1)
        particles[p, pos_dim:] = deltas.reshape(-1)
        velocities[p] = np.random.normal(scale=0.5, size=dim)
        cost = objective_timeopt(particles[p])
        pbest[p] = particles[p].copy()
        pbest_cost[p] = cost
        if cost < gbest_cost:
            gbest_cost = cost
            gbest = particles[p].copy()

    history = []
    for it in range(iters):
        r1 = np.random.rand(n_particles, dim)
        r2 = np.random.rand(n_particles, dim)
        velocities = w * velocities + c1 * r1 * (pbest - particles) + c2 * r2 * (gbest - particles)
        particles = particles + velocities
        # enforce bounds: clamp positions and deltas
        for p in range(n_particles):
            traj = particles[p, :pos_dim].reshape((n_uavs, T, 3))
            deltas = particles[p, pos_dim:].reshape((n_uavs, T-1))
            traj = clamp_pos(traj)
            # clamp deltas to positive reasonable range
            deltas = np.clip(deltas, 0.05, 10.0)
            # re-pack
            particles[p, :pos_dim] = traj.reshape(-1)
            particles[p, pos_dim:] = deltas.reshape(-1)
            cost = objective_timeopt(particles[p])
            if cost < pbest_cost[p]:
                pbest_cost[p] = cost
                pbest[p] = particles[p].copy()
            if cost < gbest_cost:
                gbest_cost = cost
                gbest = particles[p].copy()
        history.append(gbest_cost)
        if (it+1) % 25 == 0 or it == 0:
            print(f"Iter {it+1}/{iters}, best cost={gbest_cost:.3f}")

    best_pos = gbest[:pos_dim].reshape((n_uavs, T, 3))
    best_deltas = gbest[pos_dim:].reshape((n_uavs, T-1))
    return best_pos, best_deltas, history

# ----- Utilities: export and validation -----

def trajectories_to_4d_timeopt(best_traj, deltas):
    times = cumulative_times_from_deltas(deltas)
    n, T, _ = best_traj.shape
    out = np.zeros((n, T, 4))
    out[:, :, :3] = best_traj
    out[:, :, 3] = times
    return out


def validate_trajectory_timeopt(best_traj, deltas):
    """Devuelve métricas útiles: max_speed_obs, min_sep_obs, out_of_bounds_count"""
    metrics = {}
    # speeds
    dists = np.linalg.norm(np.diff(best_traj, axis=1), axis=2)  # (n_uavs,T-1)
    speeds = dists / deltas
    metrics['max_speed_obs'] = float(np.nanmax(speeds))
    # min separation (sample each UAV's times)
    times = cumulative_times_from_deltas(deltas)
    min_sep_obs = float(np.inf)
    for u in range(n_uavs):
        for i in range(T):
            p_u = best_traj[u, i]
            t_u = times[u, i]
            for v in range(u+1, n_uavs):
                p_v = interp_pos_at_time(best_traj[v], times[v], t_u)
                d = np.linalg.norm(p_u - p_v)
                if d < min_sep_obs:
                    min_sep_obs = d
    metrics['min_sep_obs'] = float(min_sep_obs)
    # bounds
    out_of_bounds = 0
    for dim in range(3):
        low = space_bounds[dim,0]
        high = space_bounds[dim,1]
        out_of_bounds += int(np.sum((best_traj[...,dim] < low) | (best_traj[...,dim] > high)))
    metrics['out_of_bounds_count'] = int(out_of_bounds)
    return metrics

# ----- Simple CSV save for timeopt trajectories -----

def save_trajectories_csv_timeopt(best_traj, deltas, filename_prefix='uav_traj_timeopt'):
    os.makedirs('output', exist_ok=True)
    traj4 = trajectories_to_4d_timeopt(best_traj, deltas)
    for i in range(n_uavs):
        data = traj4[i]
        fname = os.path.join('output', f"{filename_prefix}_uav{i}.csv")
        # columns: t,x,y,z
        data_out = np.zeros((T, 4))
        data_out[:,0] = data[:,3]
        data_out[:,1:] = data[:,0:3]
        np.savetxt(fname, data_out, delimiter=',', header='t,x,y,z', comments='')
        print(f"Guardado: {fname}")


### --- Visualización para time-optimized trajectories ---


def sample_trajectories_on_grid(best_traj, deltas, num_samples=None):
    """Muestra las trayectorias de todos los UAVs en una rejilla temporal común.
    Devuelve: times_grid (M,), samples (n_uavs, M, 3)
    Si num_samples es None se usa la unión de todos los tiempos discretos.
    """
    times = cumulative_times_from_deltas(deltas)  # (n_uavs, T)
    if num_samples is None:
        # unión de todos los tiempos (ordenados)
        grid = np.unique(times.flatten())
    else:
        tmax = np.max(times[:, -1])
        grid = np.linspace(0.0, tmax, num_samples)
    n = best_traj.shape[0]
    M = grid.size
    samples = np.zeros((n, M, 3))
    for u in range(n):
        for i, tg in enumerate(grid):
            samples[u, i] = interp_pos_at_time(best_traj[u], times[u], tg)
    return grid, samples


def plot_timeopt_results(best_traj, deltas, history=None, show=True, save_prefix=None, sample_num=None):
    """Dibuja:
    - 3D: trayectorias x,y,z con color por tiempo
    - distancias par-a-par en la rejilla temporal común
    - historial de coste (si se pasa)
    Si save_prefix se proporciona, guarda PNG en output/{save_prefix}_*.png
    """
    os.makedirs('output', exist_ok=True)
    # 3D plot
    fig = plt.figure(figsize=(10, 6))
    ax = fig.add_subplot(111, projection='3d')
    times = cumulative_times_from_deltas(deltas)
    cmap = cm.get_cmap('viridis')
    tmin = 0.0
    tmax = np.max(times[:, -1])
    for u in range(best_traj.shape[0]):
        t_u = times[u]
        traj = best_traj[u]
        # color by normalized time per point
        norm_t = (t_u - tmin) / (tmax - tmin + 1e-9)
        for i in range(traj.shape[0]-1):
            col = cmap(norm_t[i])
            ax.plot(traj[i:i+2,0], traj[i:i+2,1], traj[i:i+2,2], color=col)
        ax.scatter(traj[:,0], traj[:,1], traj[:,2], c=norm_t, cmap='viridis', s=20)
        ax.scatter(traj[0,0], traj[0,1], traj[0,2], marker='o', s=40)
        ax.scatter(traj[-1,0], traj[-1,1], traj[-1,2], marker='^', s=40)
    ax.set_xlim(space_bounds[0])
    ax.set_ylim(space_bounds[1])
    ax.set_zlim(space_bounds[2])
    ax.set_title('Trayectorias optimizadas (x,y,z) con color por tiempo')
    if save_prefix:
        fig.savefig(os.path.join('output', f"{save_prefix}_3d.png"))
    if show:
        plt.show()
    plt.close(fig)

    # Pairwise distances on a common grid
    grid, samples = sample_trajectories_on_grid(best_traj, deltas, num_samples=sample_num)
    pairs = []
    n = samples.shape[0]
    for i in range(n):
        for j in range(i+1, n):
            pairs.append((i, j))
    fig2, ax2 = plt.subplots(figsize=(10,4))
    for (i,j) in pairs:
        dists = np.linalg.norm(samples[i] - samples[j], axis=1)
        ax2.plot(grid, dists, label=f"{i}-{j}")
    ax2.axhline(min_sep, linestyle='--', color='k', label='min_sep')
    ax2.set_xlabel('Tiempo (s)')
    ax2.set_ylabel('Distancia inter-UAV')
    ax2.set_title('Distancias par-a-par sobre la rejilla temporal común')
    ax2.legend()
    if save_prefix:
        fig2.savefig(os.path.join('output', f"{save_prefix}_pairwise.png"))
    if show:
        plt.show()
    plt.close(fig2)

    # Cost history
    if history is not None:
        fig3, ax3 = plt.subplots(figsize=(8,3))
        ax3.plot(np.arange(len(history)), history)
        ax3.set_xlabel('Iteración')
        ax3.set_ylabel('Coste (gbest)')
        ax3.set_title('Evolución del coste')
        if save_prefix:
            fig3.savefig(os.path.join('output', f"{save_prefix}_cost.png"))
        if show:
            plt.show()
        plt.close(fig3)


# ----- If run as script, run a short example for demonstration -----
if __name__ == '__main__':
    # Ejecutar demo con parámetros reducidos para que termine rápido
    best_pos, best_deltas, history = run_pso_timeopt(n_particles=300, iters=500, seed=42)
    metrics = validate_trajectory_timeopt(best_pos, best_deltas)
    print('Metrics:', metrics)
    # guardar CSV de ejemplo
    save_trajectories_csv_timeopt(best_pos, best_deltas)

    # Mostrar y/o guardar gráficas. Guardamos siempre en output/ y solo mostramos si hay DISPLAY.
    has_display = bool(os.environ.get('DISPLAY'))
    if not has_display:
        print('No DISPLAY detected: guardando gráficas en output/ (modo headless)')
    plot_timeopt_results(best_pos, best_deltas, history=history, show=has_display, save_prefix='timeopt_demo', sample_num=200)

File: ros2_ws/src/px4_ros_com/src/examples/tests/4d_trajectory_timeopt.cpp
#include <bits/stdc++.h>
using namespace std;

// Simple C++ implementation of the time-optimizing PSO from the Python prototype.
// Outputs per-UAV CSV files with columns t,x,y,z and a history CSV with gbest cost.

// NOTE: This is a straightforward, single-threaded implementation intended for
// correctness and portability, not raw performance. It avoids external deps.

struct Config {
    int n_uavs = 3;
    int T = 10;
    double dt_base = 1.0;
    double max_speed = 3.0;
    double min_sep = 2.0;
    array<pair<double,double>,3> space_bounds = { make_pair(0.0,40.0), make_pair(0.0,40.0), make_pair(0.5,10.0) };
    //vector<vector<double>> starts{{2,2,3},{2,8,3},{2,14,3}};
    //vector<vector<double>> goals{{35,35,4},{35,30,6},{35,25,3}};
    vector<vector<double>> starts{{0,0,5},{0,20,5},{20,0,5}};
    vector<vector<double>> goals{{20,20,5},{20,0,5},{0,20,5}};
};

// Note: random helpers are created per-run inside main using a single rng to
// ensure reproducibility and parity with the Python prototype (seeded RNG).

using Vec = vector<double>;

inline void clamp_pos(vector<vector<vector<double>>> &traj, const Config &cfg){
    int n = cfg.n_uavs;
    int T = cfg.T;
    for(int u=0; u<n; ++u) for(int t=0;t<T;++t) for(int d=0; d<3; ++d){
        double low = cfg.space_bounds[d].first;
        double high = cfg.space_bounds[d].second;
        traj[u][t][d] = min(max(traj[u][t][d], low), high);
    }
}

// cumulative times from deltas
vector<vector<double>> cumulative_times_from_deltas(const vector<vector<double>> &deltas){
    int n = deltas.size();
    int m = deltas[0].size();
    vector<vector<double>> times(n, vector<double>(m+1, 0.0));
    for(int i=0;i<n;++i){
        for(int k=0;k<m;++k) times[i][k+1] = times[i][k] + deltas[i][k];
    }
    return times;
}

// linear interpolation for 1D index
vector<double> interp_pos_at_time(const vector<vector<double>> &traj, const vector<double> &times, double t_query){
    int T = (int)traj.size();
    if(t_query <= times[0]) return traj[0];
    if(t_query >= times.back()) return traj.back();
    int idx = (int)(upper_bound(times.begin(), times.end(), t_query) - times.begin()) - 1;
    double t0 = times[idx]; double t1 = times[idx+1];
    const vector<double> &p0 = traj[idx];
    const vector<double> &p1 = traj[idx+1];
    double alpha = (t_query - t0) / (t1 - t0 + 1e-12);
    vector<double> out(3);
    for(int d=0; d<3; ++d) out[d] = (1-alpha)*p0[d] + alpha*p1[d];
    return out;
}

// cost terms

// Distancia mínima entre dos segmentos en 3D
double segment_min_dist(const vector<double> &p1, const vector<double> &p2, const vector<double> &q1, const vector<double> &q2) {
    // Algoritmo estándar: https://stackoverflow.com/a/14772574
    vector<double> u(3), v(3), w0(3);
    for(int d=0; d<3; ++d) {
        u[d] = p2[d] - p1[d];
        v[d] = q2[d] - q1[d];
        w0[d] = p1[d] - q1[d];
    }
    double a = u[0]*u[0] + u[1]*u[1] + u[2]*u[2];
    double b = u[0]*v[0] + u[1]*v[1] + u[2]*v[2];
    double c = v[0]*v[0] + v[1]*v[1] + v[2]*v[2];
    double d = u[0]*w0[0] + u[1]*w0[1] + u[2]*w0[2];
    double e = v[0]*w0[0] + v[1]*w0[1] + v[2]*w0[2];
    double D = a*c - b*b;
    double sc, tc;
    if(D < 1e-12) {
        sc = 0.0;
        tc = (c > 1e-12) ? (e/c) : 0.0;
    } else {
        sc = (b*e - c*d) / D;
        tc = (a*e - b*d) / D;
    }
    sc = std::min<double>(std::max<double>(sc, 0.0), 1.0);
    tc = std::min<double>(std::max<double>(tc, 0.0), 1.0);
    vector<double> pt1(3), pt2(3);
    for(int d=0; d<3; ++d) {
        pt1[d] = p1[d] + sc*u[d];
        pt2[d] = q1[d] + tc*v[d];
    }
    double dist = sqrt((pt1[0]-pt2[0])*(pt1[0]-pt2[0]) + (pt1[1]-pt2[1])*(pt1[1]-pt2[1]) + (pt1[2]-pt2[2])*(pt1[2]-pt2[2]));
    return dist;
}

double path_length(const vector<vector<vector<double>>> &trajectories){
    double L = 0.0;
    for(size_t u=0; u<trajectories.size(); ++u){
        for(size_t t=1; t<trajectories[u].size(); ++t){
            double s=0; for(int d=0; d<3; ++d){ double diff = trajectories[u][t][d] - trajectories[u][t-1][d]; s += diff*diff; }
            L += sqrt(s);
        }
    }
    return L;
}

double speed_penalty_timeopt(const vector<vector<vector<double>>> &trajectories, const vector<vector<double>> &deltas, const Config &cfg){
    int n = cfg.n_uavs; int T = cfg.T;
    double pen = 0.0;
    for(int u=0; u<n; ++u){
        for(int t=0; t<T-1; ++t){
            double s=0; for(int d=0; d<3; ++d){ double diff = trajectories[u][t+1][d]-trajectories[u][t][d]; s += diff*diff; }
            double dist = sqrt(s);
            double delta = max(deltas[u][t], 1e-6);
            double speed = dist / delta;
            double excess = max(0.0, speed - cfg.max_speed);
            pen += excess*excess * 100.0;
            if(delta < 0.05) pen += (0.05 - delta)*(0.05 - delta) * 1000.0;
        }
    }
    return pen;
}

double collision_penalty_timeopt(const vector<vector<vector<double>>> &trajectories, const vector<vector<double>> &deltas, const Config &cfg){
    // Penalización por colisión en puntos discretos (como antes)
    auto times = cumulative_times_from_deltas(deltas);
    int n = cfg.n_uavs; int T = cfg.T;
    double pen = 0.0;
    for(int u=0; u<n; ++u){
        for(int i=0; i<T; ++i){
            double t_u = times[u][i];
            const vector<double> &p_u = trajectories[u][i];
            for(int v=u+1; v<n; ++v){
                vector<double> p_v = interp_pos_at_time(trajectories[v], times[v], t_u);
                double s=0; for(int d=0; d<3; ++d){ double diff = p_u[d] - p_v[d]; s += diff*diff; }
                double dist = sqrt(s);
                if(dist < cfg.min_sep) pen += (cfg.min_sep - dist)*(cfg.min_sep - dist) * 2000.0;
            }
        }
    }
    // Penalización por colisión en los segmentos entre puntos
    for(int seg_idx=0; seg_idx<T-1; ++seg_idx){
        for(int i1=0; i1<n; ++i1){
            const vector<double> &p1 = trajectories[i1][seg_idx];
            const vector<double> &p2 = trajectories[i1][seg_idx+1];
            for(int i2=i1+1; i2<n; ++i2){
                const vector<double> &q1 = trajectories[i2][seg_idx];
                const vector<double> &q2 = trajectories[i2][seg_idx+1];
                double dseg = segment_min_dist(p1, p2, q1, q2);
                if(dseg < cfg.min_sep) pen += (cfg.min_sep - dseg)*(cfg.min_sep - dseg) * 2000.0;
            }
        }
    }
    return pen;
}


double boundary_penalty(const vector<vector<vector<double>>> &trajectories, const Config &cfg){
    double pen = 0.0;
    int n = cfg.n_uavs; int T = cfg.T;
    for(int u=0; u<n; ++u) for(int t=0;t<T;++t) for(int d=0; d<3; ++d){
        double low = cfg.space_bounds[d].first; double high = cfg.space_bounds[d].second;
        double v = trajectories[u][t][d];
        if(v < low) pen += (low - v)*(low - v) * 100.0;
        if(v > high) pen += (v - high)*(v - high) * 100.0;
    }
    return pen;
}


double objective_timeopt_flat(const Vec &flat, const Config &cfg){
    int pos_len = cfg.n_uavs * cfg.T * 3;
    int del_len = cfg.n_uavs * (cfg.T-1);
    // unpack
    vector<vector<vector<double>>> traj(cfg.n_uavs, vector<vector<double>>(cfg.T, vector<double>(3)));
    vector<vector<double>> deltas(cfg.n_uavs, vector<double>(cfg.T-1));
    int idx=0;
    for(int u=0; u<cfg.n_uavs; ++u) for(int t=0;t<cfg.T;++t) for(int d=0;d<3;++d) traj[u][t][d] = flat[idx++];
    for(int u=0; u<cfg.n_uavs; ++u) for(int t=0;t<cfg.T-1;++t) deltas[u][t] = flat[idx++];
    // penalties
    double start_err=0, goal_err=0;
    for(int u=0; u<cfg.n_uavs; ++u){ for(int d=0; d<3; ++d){ double v0 = traj[u][0][d] - cfg.starts[u][d]; start_err += v0*v0; double v1 = traj[u][cfg.T-1][d] - cfg.goals[u][d]; goal_err += v1*v1; } }
    start_err *= 1000.0; goal_err *= 1000.0;
    double L = path_length(traj);
    double sp = speed_penalty_timeopt(traj, deltas, cfg);
    double coll = collision_penalty_timeopt(traj, deltas, cfg);
    double bnd = boundary_penalty(traj, cfg);
    double pos_smooth = 0.0; // simple second-diff smoothness
    for(int u=0; u<cfg.n_uavs; ++u){
        for(int t=0; t<cfg.T-2; ++t){ double s=0; for(int d=0;d<3;++d){ double a=traj[u][t+2][d]-2*traj[u][t+1][d]+traj[u][t][d]; s+=a*a; } pos_smooth += sqrt(s); }
    }
    pos_smooth *= 500.0;
    // time smoothness approximate: sum abs diff of deltas
    double time_smooth = 0.0;
    for(int u=0; u<cfg.n_uavs; ++u) for(int t=0; t<cfg.T-2; ++t) time_smooth += fabs(deltas[u][t+1]-deltas[u][t]);
    time_smooth *= 50.0;
    double cost = L + pos_smooth + time_smooth + sp + coll + bnd + start_err + goal_err;
    return cost;
}

// Save CSVs
void save_trajectories_csv(const vector<vector<vector<double>>> &traj, const vector<vector<double>> &deltas, const string &prefix="uav_traj_cpp"){
    int n = traj.size(); int T = traj[0].size();
    auto times = cumulative_times_from_deltas(deltas);
    for(int u=0; u<n; ++u){
        string fname = string("output/") + prefix + "_uav" + to_string(u) + ".csv";
        ofstream f(fname);
        f << "t,x,y,z\n";
        for(int t=0;t<T;++t){
            double tt = times[u][t];
            f << tt << "," << traj[u][t][0] << "," << traj[u][t][1] << "," << traj[u][t][2] << "\n";
        }
        f.close();
    }
}

int main(int argc, char** argv){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    Config cfg;
    // PSO params — align with Python prototype defaults
    int n_particles = 200;
    int iters = 500;
    double w = 0.6, c1 = 1.8, c2 = 1.8;
    // seed
    std::mt19937_64 rng(42);
    // distributions and convenience lambdas (use same rng for reproducibility)
    std::normal_distribution<double> norm01(0.0, 1.0);
    std::uniform_real_distribution<double> uni01(0.0, 1.0);
    auto randn = [&](double mean=0.0, double std=1.0){ return mean + std * norm01(rng); };
    auto randu = [&](){ return uni01(rng); };

    int pos_dim = cfg.n_uavs * cfg.T * 3;
    int del_dim = cfg.n_uavs * (cfg.T-1);
    int dim = pos_dim + del_dim;

    vector<Vec> particles(n_particles, Vec(dim, 0.0));
    vector<Vec> velocities(n_particles, Vec(dim, 0.0));
    vector<Vec> pbest(n_particles, Vec(dim, 0.0));
    vector<double> pbest_cost(n_particles, 1e300);
    Vec gbest(dim, 0.0); double gbest_cost = 1e300;

    // init
    for(int p=0;p<n_particles;++p){
        // straight-line traj + noise
        int idx=0;
        for(int u=0; u<cfg.n_uavs; ++u){
            for(int t=0;t<cfg.T;++t){
                for(int d=0; d<3; ++d){
                    double v = cfg.starts[u][d] + (cfg.goals[u][d]-cfg.starts[u][d]) * (double)t / (cfg.T-1);
                    v += randn(0.0, 0.6);
                    particles[p][idx++] = v;
                }
            }
        }
        for(int u=0; u<cfg.n_uavs; ++u) for(int t=0;t<cfg.T-1;++t) particles[p][idx++] = max(0.05, cfg.dt_base + randn(0.0, 0.1));
        for(int i=0;i<dim;++i) velocities[p][i] = randn(0.0, 0.5);
        double cost = objective_timeopt_flat(particles[p], cfg);
        pbest[p] = particles[p]; pbest_cost[p] = cost;
        if(cost < gbest_cost){ gbest_cost = cost; gbest = particles[p]; }
    }

    vector<double> history;
    history.reserve(iters);

    for(int it=0; it<iters; ++it){
        for(int p=0;p<n_particles;++p){
            for(int k=0;k<dim;++k){
                double r1 = randu(); double r2 = randu();
                velocities[p][k] = w * velocities[p][k] + c1*r1*(pbest[p][k] - particles[p][k]) + c2*r2*(gbest[k] - particles[p][k]);
                particles[p][k] += velocities[p][k];
            }
            // clamp positions and deltas
            int idx=0;
            for(int u=0; u<cfg.n_uavs; ++u){
                for(int t=0;t<cfg.T;++t){
                    for(int d=0; d<3; ++d){
                        double low = cfg.space_bounds[d].first, high = cfg.space_bounds[d].second;
                        particles[p][idx] = min(max(particles[p][idx], low), high);
                        ++idx;
                    }
                }
            }
            for(int u=0; u<cfg.n_uavs; ++u) for(int t=0;t<cfg.T-1;++t){ particles[p][idx] = min(max(particles[p][idx], 0.05), 10.0); ++idx; }

            double cost = objective_timeopt_flat(particles[p], cfg);
            if(cost < pbest_cost[p]){ pbest_cost[p] = cost; pbest[p] = particles[p]; }
            if(cost < gbest_cost){ gbest_cost = cost; gbest = particles[p]; }
        }
        history.push_back(gbest_cost);
        if((it+1) % 25 == 0 || it==0) cout << "Iter " << (it+1) << "/" << iters << ", best cost=" << gbest_cost << "\n";
    }

    // unpack gbest
    vector<vector<vector<double>>> best_traj(cfg.n_uavs, vector<vector<double>>(cfg.T, vector<double>(3)));
    vector<vector<double>> best_deltas(cfg.n_uavs, vector<double>(cfg.T-1));
    int idx=0;
    for(int u=0; u<cfg.n_uavs; ++u) for(int t=0;t<cfg.T;++t) for(int d=0; d<3; ++d) best_traj[u][t][d] = gbest[idx++];
    for(int u=0; u<cfg.n_uavs; ++u) for(int t=0;t<cfg.T-1;++t) best_deltas[u][t] = gbest[idx++];

    // ensure output dir
    system("mkdir -p output");
    save_trajectories_csv(best_traj, best_deltas, "uav_traj_cpp");
    // save history
    ofstream fh("output/history_cpp.csv"); fh << "iter,cost\n"; for(size_t i=0;i<history.size();++i) fh << i << "," << history[i] << "\n"; fh.close();
    cout << "Saved CSVs in output/" << endl;
    return 0;
}
